src/common/datomish/transact.cljc

```diff
@@ -322,6 +322,221 @@
+    (util/group-by-kv evolve1 allocations-v)))
+
+(defn evolve-allocations-ev [id->e allocations-ev]
+  "Given a map id->e of id-literals to integer entids, evolve the entities in allocations-ev.  Returns a
+  map whose keys are generations and whose values are vectors of entities in those generations."
+  (let [evolve1
+        (fn [[op id-e a id-v]]
+          (let [e* (get id->e id-e)
+                v* (get id->e id-v)]
+            (if e*
+              (if v*
+                [:resolved [op e* a v*]]
+                [:allocations-v [op e* a id-v]])
+              (if v*
+                [:allocations-e [op id-e a v*]]
+                [:allocations-ev [op id-e a id-v]]))))]
+    (util/group-by-kv evolve1 allocations-ev)))
+
+(defn <evolve [db evolution]
+  "Evolve a map of generations {:upserts-e [...], :upserts-ev [...], ...} as much as possible.
+
+  The algorithm is as follows.
+
+  First, resolve :upserts-e against the database.  Some [a v] -> e will upsert; some will not.
+  Some :upserts-e evolve to become actual :upserts (they upserted!); any other :upserts-e evolve to
+  become :allocations-e (they did not upsert, and will not upsert this transaction).
+
+  Using the newly upserted id-literals, some :upserts-ev evolve to become :resolved;
+  some :upserts-ev evolve to become :upserts-e; and some :upserts-ev remain :upserts-ev.
+
+  Likewise, some :allocations-ev evolve to become :allocations-e, :allocations-v, or :resolved; some
+  :allocations-e evolve to become :resolved; and some :allocations-v evolve to become :resolved.
+
+  If we have *new* :upserts-e (i.e., some :upserts-ev become :upserts-e), then we may be able to
+  make more progress.  We recurse, trying to resolve these new :upserts-e.
+
+  Eventually we will have no :upserts-e.  At this point, :upserts-ev become :allocations-ev, and now
+  we have :entities, :upserted, :resolved, and various :allocations-*.
+
+  As a future optimization, :upserts do not need to be inserted; they upserted, so they already
+  exist in the DB.  (We still need to verify uniqueness and ensure no overlapping can occur.)
+  Similary, :allocations-* do not need to be checked for existence, so they can be written to the DB
+  faster."
+  (go-pair
+    (let [{:keys [upserted resolved upserts-ev upserts-e allocations-ev allocations-e allocations-v entities]} evolution]
+      (if-not upserts-e
+        ;; No more progress to be made.  Any upserts-ev must just be allocations.
+        {:allocations-ev (concat upserts-ev allocations-ev)
+         :allocations-e allocations-e
+         :allocations-v allocations-v
+         :upserted upserted
+         :resolved resolved
+         :entities entities}
+        ;; Progress can be made.  Try to evolve further.
+        (let [id->e (<? (<resolve-upserts-e db upserts-e))]
+          (merge-with
+            concat
+            {:tempids  id->e ;; TODO: ensure we handle conflicting upserts across generations correctly here.
```

Comment:
Yeah, I was wondering this. What if in the first pass something's an `ev` waiting, and a `v` known? The unknown `v` gets resolved to an ID, but the `v` is known, so we push that tuple into `resolved`. In the next pass we now know the `v`, so we have an `upserts-e`. That resolves to a different entity, but we don't know there's a conflict because the other was already resolved.

So I think you need to keep `id->e` mappings around across generations.